tools/clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -337,26 +338,14 @@ void EmitAssemblyHelper::CreatePasses() {
   PMBuilder.HLSLResMayAlias = CodeGenOpts.HLSLResMayAlias;
   PMBuilder.ScanLimit = CodeGenOpts.ScanLimit;
 
-  PMBuilder.EnableGVN = !CodeGenOpts.HLSLOptimizationToggles.count("gvn") ||
-                        CodeGenOpts.HLSLOptimizationToggles.find("gvn")->second;
-
-  PMBuilder.HLSLNoSink = CodeGenOpts.HLSLOptimizationToggles.count("sink") &&
-                         !CodeGenOpts.HLSLOptimizationToggles.find("sink")->second;
-
-  PMBuilder.StructurizeLoopExitsForUnroll =
-                        !CodeGenOpts.HLSLOptimizationToggles.count("structurize-loop-exits-for-unroll") ||
-                        CodeGenOpts.HLSLOptimizationToggles.find("structurize-loop-exits-for-unroll")->second;
-
-  PMBuilder.HLSLEnableDebugNops =
-                        !CodeGenOpts.HLSLOptimizationToggles.count("debug-nops") ||
-                        CodeGenOpts.HLSLOptimizationToggles.find("debug-nops")->second;
-
-  PMBuilder.HLSLEnableLifetimeMarkers =
-      CodeGenOpts.HLSLEnableLifetimeMarkers &&
-      (!CodeGenOpts.HLSLOptimizationToggles.count("lifetime-markers") ||
-       CodeGenOpts.HLSLOptimizationToggles.find("lifetime-markers")->second);
-
-  PMBuilder.HLSLEnablePartialLifetimeMarkers = CodeGenOpts.HLSLEnablePartialLifetimeMarkers;
+  if (const std::shared_ptr<hlsl::HLSLExtensionsCodegenHelper> &extHelper = CodeGenOpts.HLSLExtensionsCodegen) {
```

Comment:
If `HLSLExtensionsCodegenHelper` is not setup, then the opt-enable stuff also didn't get setup. So we do require it here. I'll add an assert.

Reply:
The one thing that's standing out to me with this change is that it seems to put a lot more emphasis on having the extensions helper there to generate code correctly. I'm not sure if that is a big change or not (i.e. did we always have it available anyway).

In some ways it feels like we are tying two independent things together by placing the optimization toggles there. It feels like there should be some way to setup and query the optimization toggles. Then if we have extensions present, we use those to modify the current optimization toggles.

Would it make sense to separate these things more do you think?

Reply:
The reason everything is in `HLSLExtensionsCodegenHelper` is a single instance shared by the many copies of CodeGenOpts. There are many different copies of CodeGenOpts throughout the frontend; in the old code we only modified the original copy in `CompilerInstance` and the copy in `CodeGen::CodeGenModule` in `CodeGeneratorImpl::HandleTranslationUnit`, which successfully affected the `PassManagerBuilder`, but this was both too late and not the right copy to affect the earlier codegen.

If we stick to modifying the CodeGenOpts, but just doing it earlier, we will need to find the right copy to modify (which is not easy to find the right one (because there's many), and inconvenient because they're hidden inside abstract 'handler', 'action' classes, and kind hacky (becaues they're all supposed to be const)).

To sum up: I think to make 'reading semantic define earlier` work, relying on `HLSLExtensionsCodegenHelper` is the cleanest way to do it.

(Actually, an even cleaner way to do it would be to preprocess the input first before we even setup the original first copy of CodeGenOpts. However, I think that's a much much more drastic, riskier change.)

Reply:
What if we separated out the optimization toggles into its own shared_ptr that is shared by all the `CodeGenOpts` instances? It sounds a bit like we are putting them in the `HLSLExtensionsCodegenHelper` because it is a single place that is common to all the `CodeGenOpts` copies, but otherwise it doesn't really belong there.

Could we add the `IsOptionEnabled()` function to the new `OptToggles` class instead of putting it in `HLSLExtensionsCodegenHelper`? Then we can check whether the options are enabled without requiring the extensions. After parsing semantic defines the helper could update the shared `OptToggles` state with the values coming from the defines.

Reply:
I like this suggestion. Conceptually I didn't register that opt-enable/disable toggles is a separate concept from the extension stuff. Updated in a new push.